# STA130 Homework Week 10
ChatGPT URL: https://chatgpt.com/share/6736a858-dff4-8002-bc86-f0cea6576a4b

1.1) The difference between simple linear regression and multiple linear regression is dependent on the number of variables/predictors they each take. In the case of single linear regression, the formula is $$y = \beta_0 + \beta_1 x + \epsilon$$, where x is the predictor and y is the variable. However, in the case of multiple linear regression, it takes multiple predictor variables to calculate a single output. The formula for multiple linear regression is $$y = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + \dots + \beta_n x_n + \epsilon$$. Independent/predictor variables are the variables we associate with change, for example, in the case of the housing market, the variable y is the output, and can be represented by something like the housing price. A variable x could be the age of the house, and potentially in the case of multiple linear regresison, there could be additional variables such as the neighbourhood safety, city, and other pieces of data. As directly mentioned before, multiple linear regression holds more benefits over single linear regression as it gives us more control over the variables and as such, we can determine more information using more variables, isolating causes for potential rises and decreasing the chance for errors.

1.2) As shown in last week's homework with the water eruptions, regarding the continuous variable and indicator variable in parameter testing, a continuous variable is akin to a histogram covering each individual point. There are no bins and the data is ungrouped. However, in an indicator variable testing, we are able to convert the data to a binary value which lets us determine the difference between two things if we're roughly sure about what could cause a change. Using the example from last week, with water geyser eruptions, we were roughly able to determine that the thing having an effect was the duration and as such, we were able to split the data into two bins of "short" vs "long' waits.

1.3) A continuous variable, as mentioned above, shows individual differences, or almost a "minute by minute" or "item by item" breakdown of the thing being tested as the perdictor variable. However, the indicator variable groups that into large groups, and as such, you could see the effect of a certain variable and its effects on the output variable. This could be beneficial as it would change the data in accordance with now having two sources of input, in that it can capture things specific to a certain group of data, whilst also accounting for the minute by minute or item by item breakdowns.

1.4) Assuming this is relatively similar to what was discussed in Question 1.3, the graph will now have a slight slope dependent on the group. ChatGPT provided an example of the impact of income and life satisfaction between those with vs without a college degree. This would let us determine if income has a larger effect on life satisfaction on those with a degree, vs those without a degree.

1.5) This question utilized a lot of big words, but the way I understand it is that you are converting a non binary categorical variable (something that is likely continuous) into a set of indicator variables. In this case, we can categorize the data into many groups which functionally would work as a lot of indicator variables. We could see the effects of various "bins" of people within a certain category. As an example, we might set the outcome variable to be the housing price, then set a categorical variable as something like house location (downtowns, suburbs, countryside). The indicators could be "1 if in suburbs, 0 otherwise) and (1 if in downtown, 0 otherwise). In this example we would have the baseline housing price be the price in the countryside, and anywhere where the bin = 1 for each of the indicator variables be for its respective location. 

2.1) In the case of the question, the outcome variable (y) would likely be something like the revenue/sales generated by the ad campaign. The predictor variables would be the money spent on online advertising and the money spent on TV advertising (which is 100% less effective with a near guarantee. this is 2024). After speaking with ChatGPT, there are multiple potential interaction effects but one provided by ChatGPT was discussing the impact of TV ads after already having online ads. Having reinforcement on both TV and online could result in a more effective ad campaign. The equations used would be the same as the multiple lienar regression method, with the one involving an interaction also adding on a third variable being TV spend + online spend or any other interaction between the two. We can use the formulas to make predictions about the outcome by simply plugging in values, for example in the case of the formula involving interaction, indicated by $$\text{Sales} = \beta_0 + \beta_1 (\text{TV Spend}) + \beta_2 (\text{Online Spend}) + \beta_3 (\text{TV Spend} \times \text{Online Spend}) + \epsilon$$, we could choose two values of how much you would potentially spend on TV, and how much you would potentially spend on online advertising. You would then multiply by a previously determined coefficient, in these examples, beta. As a result, you are simply plugging in numbers. in the case of the differences between the formulas with and without interaction, it is exactly as it states. In one, you add on the possibility for interaction between the two forms of advertising and its effects on your campaigns, and in the other, you simply assume they are completely independent with little to no overlap.

2.2) We would simply be converting these into two indicator variables, letting ITV be 1 if it is high spending and 0 if it is low spending. We then use the same formulas as above, but simply changing the TV spend and online spend variables to be the indicators respectively. The outcome will likely be roughly the same but just less detailed, simply focusing on the "high vs low" spending.

In [32]:
import pandas as pd
import statsmodels.formula.api as smf

url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
pokeaman = pd.read_csv(url).fillna('None')

pokeaman['str8fyre'] = (pokeaman['Type 1']=='Fire').astype(int)
linear_model_specification_formula = \
'str8fyre ~ Attack*Legendary + Defense*I(Q("Type 2")=="None") + C(Generation)'
log_reg_fit = smf.logit(linear_model_specification_formula, data=pokeaman).fit()
log_reg_fit.summary()

Optimization terminated successfully.
         Current function value: 0.228109
         Iterations 8


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               str8fyre   No. Observations:                  800
Model:                          Logit   Df Residuals:                      788
Method:                           MLE   Df Model:                           11
Date:                Fri, 15 Nov 2024   Pseudo R-squ.:                 0.05156
Time:                        03:28:41   Log-Likelihood:                -182.49
converged:                       True   LL-Null:                       -192.41
Covariance Type:            nonrobust   LLR p-value:                   0.04757
============================================================================================================
                                               coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------
Intercept                                   -3.2644      0.714     -4.572      0.000      -4.664      -1.865
Legendary[T.True]                            4.3478      2.179      1.996      0.046       0.078       8.618
I(Q("Type 2") == "None")[T.True]             1.5432      0.853      1.810      0.070      -0.128       3.215
C(Generation)[T.2]                          -0.0574      0.468     -0.123      0.902      -0.975       0.861
C(Generation)[T.3]                          -0.6480      0.466     -1.390      0.164      -1.561       0.265
C(Generation)[T.4]                          -0.8255      0.545     -1.516      0.130      -1.893       0.242
C(Generation)[T.5]                          -0.5375      0.449     -1.198      0.231      -1.417       0.342
C(Generation)[T.6]                           0.3213      0.477      0.673      0.501      -0.614       1.257
Attack                                       0.0172      0.006      3.086      0.002       0.006       0.028
Attack:Legendary[T.True]                    -0.0365      0.019     -1.884      0.060      -0.074       0.001
Defense                                     -0.0098      0.008     -1.247      0.213      -0.025       0.006
Defense:I(Q("Type 2") == "None")[T.True]    -0.0197      0.012     -1.651      0.099      -0.043       0.004
============================================================================================================
"""

3) Even with ChatGPT, frankly I have no idea how to do this. If I remember to, I will talk to you in class about this since I am interested in solving this, but even after discussing with other classmates, we don't really understand it and the github does kind of ramble a bit, so it's hard to tell what parts to actually follow with so many chunks crossed out.

$$R^2 = 1 - \frac{\sum{i=1}^n(y_i-\hat y)^2}{\sum_{i=1}^n(y_i-\bar y)^2}$$

In [33]:
import pandas as pd

url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
# fail https://github.com/KeithGalli/pandas/blob/master/pokemon_data.csv
pokeaman = pd.read_csv(url) 
pokeaman

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

In [34]:
import statsmodels.formula.api as smf

model1_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation) + Q("Sp. Def"):C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") * C(Generation)', data=pokeaman)

model2_fit = model2_spec.fit()
model2_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.176
Model:                            OLS   Adj. R-squared:                  0.164
Method:                 Least Squares   F-statistic:                     15.27
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           3.50e-27
Time:                        03:28:42   Log-Likelihood:                -3649.4
No. Observations:                 800   AIC:                             7323.
Df Residuals:                     788   BIC:                             7379.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
===================================================================================================
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                          26.8971      5.246      5.127      0.000      16.599      37.195
C(Generation)[T.2]                 20.0449      7.821      2.563      0.011       4.692      35.398
C(Generation)[T.3]                 21.3662      6.998      3.053      0.002       7.629      35.103
C(Generation)[T.4]                 31.9575      8.235      3.881      0.000      15.793      48.122
C(Generation)[T.5]                  9.4926      7.883      1.204      0.229      -5.982      24.968
C(Generation)[T.6]                 22.2693      8.709      2.557      0.011       5.173      39.366
Q("Sp. Def")                        0.5634      0.071      7.906      0.000       0.423       0.703
Q("Sp. Def"):C(Generation)[T.2]    -0.2350      0.101     -2.316      0.021      -0.434      -0.036
Q("Sp. Def"):C(Generation)[T.3]    -0.3067      0.093     -3.300      0.001      -0.489      -0.124
Q("Sp. Def"):C(Generation)[T.4]    -0.3790      0.105     -3.600      0.000      -0.586      -0.172
Q("Sp. Def"):C(Generation)[T.5]    -0.0484      0.108     -0.447      0.655      -0.261       0.164
Q("Sp. Def"):C(Generation)[T.6]    -0.3083      0.112     -2.756      0.006      -0.528      -0.089
==============================================================================
Omnibus:                      337.229   Durbin-Watson:                   1.505
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2871.522
Skew:                           1.684   Prob(JB):                         0.00
Kurtosis:                      11.649   Cond. No.                     1.40e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

4) In this equation, we see that the R squared value is 0.176, which can be converted into being 17.6%. This means that the model explains for 17.6% of the variability in the data. This could mean that there are other key variables missing from the calculation, meaning that there is missing data and potentially more variables that need to be made and related to the HP, which in this case is the dependent variable. The predictor variables of the pokeaman's special defense and their generation are combined to basically see how the effect of the special defense on the pokeaman's health differs generation to generation. The explanation for the coefficients being larger than 10 whilst having strong or very strong evidence against the null hypothesis of "no correlation" can be explained by the two numbers referencing very different things. In the case of the R-squared variable, it is solely talking about the fit of the model and not necessarily the individual significance of the predictor model. However, in the case of the coefficients being larger than 10 whilst having strong evidence against the null hypothesis, this does not directly correspond to the fit of the model and instead focuses on the individual significance of each variable. The low R squared means that whilst there is signficiant effects on the outcome, they only account for a very small variability in the graph and much of it is left unexplained. The small p values mean that the predictors are still related to the outcome, though not being the only outcome it appears.

In this conversation, we explored various aspects of a multiple linear regression model applied to data where the dependent variable, HP, is predicted using "Special Defense" (Sp. Def) and "Generation" as predictor variables. We discussed how R-squared, specifically a value of 17.6% in this model, reflects the proportion of variability in HP explained by the model. This relatively low R-squared indicates that while the predictors have some statistically significant effects on HP, they collectively account for only a modest amount of its overall variance, suggesting that other factors might influence HP or that it is highly variable.

The model includes "Special Defense" as a continuous predictor and "Generation" as a categorical variable, with dummy variables representing different generations. In the second model specification, an interaction term between "Special Defense" and "Generation" is added to allow the effect of "Special Defense" on HP to vary across generations. This interaction term is useful in scenarios where the relationship between HP and "Special Defense" might differ by generation, potentially capturing a more nuanced effect. The syntax of the model is structured in Python using statsmodels formula API, and we discussed the difference between including an interaction term (Q("Sp. Def"):C(Generation)) and using main effects alone in the model.

Overall, the addition of interactions provides insight into whether predictor effects vary by category, though the overall low R-squared suggests limited predictive power from the variables included.

(For some reason, ChatGPT will not give any summary regarding what we talked about before.)

ChatGPT URL: https://chatgpt.com/share/6736a858-dff4-8002-bc86-f0cea6576a4b

# STA130 Homework Post-Lecture
ChatGPT URL: https://chatgpt.com/share/6736bfbb-fa88-8002-a19b-1aa49629d23e

In [35]:
import numpy as np
from sklearn.model_selection import train_test_split

fifty_fifty_split_size = int(pokeaman.shape[0]*0.5)

# Replace "NaN" (in the "Type 2" column with "None")
pokeaman.fillna('None', inplace=True)

np.random.seed(130)
pokeaman_train,pokeaman_test = \
  train_test_split(pokeaman, train_size=fifty_fifty_split_size)
pokeaman_train

#                     Name   Type 1   Type 2  HP  Attack  Defense  \
370  338                  Solrock     Rock  Psychic  70      95       85   
6      6                Charizard     Fire   Flying  78      84       78   
242  224                Octillery    Water     None  75     105       75   
661  600                    Klang    Steel     None  60      80       95   
288  265                  Wurmple      Bug     None  45      45       35   
..   ...                      ...      ...      ...  ..     ...      ...   
522  471                  Glaceon      Ice     None  65      60      110   
243  225                 Delibird      Ice   Flying  45      55       45   
797  720      HoopaHoopa Confined  Psychic    Ghost  80     110       60   
117  109                  Koffing   Poison     None  40      65       95   
409  373  SalamenceMega Salamence   Dragon   Flying  95     145      130   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
370       55       65     70           3      False  
6        109       85    100           1      False  
242      105       75     45           2      False  
661       70       85     50           5      False  
288       20       30     20           3      False  
..       ...      ...    ...         ...        ...  
522      130       95     65           4      False  
243       65       45     75           2      False  
797      150      130     70           6       True  
117       60       45     35           1      False  
409      120       90    120           3      False  

[400 rows x 12 columns]

In [36]:
model_spec3 = smf.ols(formula='HP ~ Attack + Defense', 
                      data=pokeaman_train)
model3_fit = model_spec3.fit()
model3_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.148
Model:                            OLS   Adj. R-squared:                  0.143
Method:                 Least Squares   F-statistic:                     34.40
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           1.66e-14
Time:                        03:28:43   Log-Likelihood:                -1832.6
No. Observations:                 400   AIC:                             3671.
Df Residuals:                     397   BIC:                             3683.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     42.5882      3.580     11.897      0.000      35.551      49.626
Attack         0.2472      0.041      6.051      0.000       0.167       0.327
Defense        0.1001      0.045      2.201      0.028       0.011       0.190
==============================================================================
Omnibus:                      284.299   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5870.841
Skew:                           2.720   Prob(JB):                         0.00
Kurtosis:                      20.963   Cond. No.                         343.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [37]:
yhat_model3 = model3_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model3_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model3)[0,1]**2)

'In sample' R-squared:     0.14771558304519894
'Out of sample' R-squared: 0.21208501873920738


In [38]:
model4_linear_form = 'HP ~ Attack * Defense * Speed * Legendary'
model4_linear_form += ' * Q("Sp. Def") * Q("Sp. Atk")'
# DO NOT try adding '* C(Generation) * C(Q("Type 1")) * C(Q("Type 2"))'
# That's 6*18*19 = 6*18*19 possible interaction combinations...
# ...a huge number that will blow up your computer

model4_spec = smf.ols(formula=model4_linear_form, data=pokeaman_train)
model4_fit = model4_spec.fit()
model4_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.467
Model:                            OLS   Adj. R-squared:                  0.369
Method:                 Least Squares   F-statistic:                     4.764
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           4.23e-21
Time:                        03:28:43   Log-Likelihood:                -1738.6
No. Observations:                 400   AIC:                             3603.
Df Residuals:                     337   BIC:                             3855.
Df Model:                          62                                         
Covariance Type:            nonrobust                                         
====================================================================================================================================
                                                                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------------------
Intercept                                                          521.5715    130.273      4.004      0.000     265.322     777.821
Legendary[T.True]                                                   -6.1179      2.846     -2.150      0.032     -11.716      -0.520
Attack                                                              -8.1938      2.329     -3.518      0.000     -12.775      -3.612
Attack:Legendary[T.True]                                         -1224.9610    545.105     -2.247      0.025   -2297.199    -152.723
Defense                                                             -6.1989      2.174     -2.851      0.005     -10.475      -1.923
Defense:Legendary[T.True]                                         -102.4030     96.565     -1.060      0.290    -292.350      87.544
Attack:Defense                                                       0.0985      0.033      2.982      0.003       0.034       0.164
Attack:Defense:Legendary[T.True]                                    14.6361      6.267      2.336      0.020       2.310      26.963
Speed                                                               -7.2261      2.178     -3.318      0.001     -11.511      -2.942
Speed:Legendary[T.True]                                            704.8798    337.855      2.086      0.038      40.309    1369.450
Attack:Speed                                                         0.1264      0.038      3.351      0.001       0.052       0.201
Attack:Speed:Legendary[T.True]                                       5.8648      2.692      2.179      0.030       0.570      11.160
Defense:Speed                                                        0.1026      0.039      2.634      0.009       0.026       0.179
Defense:Speed:Legendary[T.True]                                     -6.9266      3.465     -1.999      0.046     -13.742      -0.111
Attack:Defense:Speed                                                -0.0016      0.001     -2.837      0.005      -0.003      -0.001
Attack:Defense:Speed:Legendary[T.True]                              -0.0743      0.030     -2.477      0.014      -0.133      -0.015
Q("Sp. Def")                                                        -5.3982      1.938     -2.785      0.006      -9.211      -1.586
Legendary[T.True]:Q("Sp. Def")                                    -282.2496    126.835     -2.225      0.027    -531.738     -32.761
Attack:Q("Sp. Def")                                                  0.1094      0.034      3.233      0.001       0.043       0.176
Attack:Legendary[T.True]:Q("Sp. Def")                               12.6503      5.851      2.162      0.031       1.141      24.160
Defense:Q("Sp. Def")    

In [39]:
yhat_model4 = model4_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model4_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model4)[0,1]**2)

'In sample' R-squared:     0.46709442115833855
'Out of sample' R-squared: 0.002485342598992873


5) The first cell of code basically prepares the dataset, splitting it in half and replacing the values in the second type's column that are missing with None.

The second cell is a multiple linear regression relating HP (the dependent variable) against attack and defense, the two independent variables. It then stores a line of best fit then summarizes that line of best fit.

The third cell calculates the R-squared values for both in sample and out of sample (the two that we split earlier), and returns the values associated. 

The fourth cell is a linear regression with more factors, such as now adding an interaction between attack, defense, speed and whether or not the pokemon is legendary. In the fifth cell, you then do the same thing the third cell does and evaluate all the information then print out the R squared values.

This code models the effect of attack and defense on HP in the first simpler linear regression, and later, relates it to almost every other factor of the pokemon in the more complex linear regression. It shows that when not accounting for the pokemon being legendary or their speed, the model can only account for just under 15% of the variation in the graph. However, when taking into consideration the pokeaman's status as a legendary vs a non legendary, as well as their speed, we can see that the R squared value is significantly higher at nearly 50%. This means that it can explain up to 50% of the variation in the graph. There is still a long way to go, but this is a signficant and noticeable improvement in how accurate the model is.

In [40]:
model3_fit.summary() 

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.148
Model:                            OLS   Adj. R-squared:                  0.143
Method:                 Least Squares   F-statistic:                     34.40
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           1.66e-14
Time:                        03:28:44   Log-Likelihood:                -1832.6
No. Observations:                 400   AIC:                             3671.
Df Residuals:                     397   BIC:                             3683.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     42.5882      3.580     11.897      0.000      35.551      49.626
Attack         0.2472      0.041      6.051      0.000       0.167       0.327
Defense        0.1001      0.045      2.201      0.028       0.011       0.190
==============================================================================
Omnibus:                      284.299   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5870.841
Skew:                           2.720   Prob(JB):                         0.00
Kurtosis:                      20.963   Cond. No.                         343.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [41]:
from patsy import center, scale

model3_linear_form_center_scale = \
  'HP ~ scale(center(Attack)) + scale(center(Defense))' 
model_spec3_center_scale = smf.ols(formula=model3_linear_form_center_scale,
                                   data=pokeaman_train)
model3_center_scale_fit = model_spec3_center_scale.fit()
model3_center_scale_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.148
Model:                            OLS   Adj. R-squared:                  0.143
Method:                 Least Squares   F-statistic:                     34.40
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           1.66e-14
Time:                        03:28:44   Log-Likelihood:                -1832.6
No. Observations:                 400   AIC:                             3671.
Df Residuals:                     397   BIC:                             3683.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                 69.3025      1.186     58.439      0.000      66.971      71.634
scale(center(Attack))      8.1099      1.340      6.051      0.000       5.475      10.745
scale(center(Defense))     2.9496      1.340      2.201      0.028       0.315       5.585
==============================================================================
Omnibus:                      284.299   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5870.841
Skew:                           2.720   Prob(JB):                         0.00
Kurtosis:                      20.963   Cond. No.                         1.66
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [42]:
model4_linear_form_CS = 'HP ~ scale(center(Attack)) * scale(center(Defense))'
model4_linear_form_CS += ' * scale(center(Speed)) * Legendary' 
model4_linear_form_CS += ' * scale(center(Q("Sp. Def"))) * scale(center(Q("Sp. Atk")))'
# Legendary is an indicator, so we don't center and scale that

model4_CS_spec = smf.ols(formula=model4_linear_form_CS, data=pokeaman_train)
model4_CS_fit = model4_CS_spec.fit()
model4_CS_fit.summary().tables[-1]  # Cond. No. is 2,250,000,000,000,000

# The condition number is still bad even after centering and scaling

<class 'statsmodels.iolib.table.SimpleTable'>

In [43]:
# Just as the condition number was very bad to start with
model4_fit.summary().tables[-1]  # Cond. No. is 12,000,000,000,000,000

<class 'statsmodels.iolib.table.SimpleTable'>

6) Model4_linear_form creates new predictor variables for Sp Def, Sp Atk, Legendary and Speed. It then takes into account the interactions between each of the variables. As a result the data is "more accurate" as to capturing the true nature of the graph, however, the data ends up becoming very multicollinear, meaning that many of the predictors are super correlated and as a result, the model cannot easily tell the difference between what each predictor does for the overall model. This could mean that the model is less accurate to the overall data, as it could be super accurate to the model but super inaccurate to new unseen data as there is simply not enough information due to how correlated and simialr some pieces of data are. In short, the model is not able to tell what one thing does and what another thing does, because to the model they seem too correlated to have their own independent differences.

In [44]:
# Here's something a little more reasonable...
model5_linear_form = 'HP ~ Attack + Defense + Speed + Legendary'
model5_linear_form += ' + Q("Sp. Def") + Q("Sp. Atk")'
model5_linear_form += ' + C(Generation) + C(Q("Type 1")) + C(Q("Type 2"))'

model5_spec = smf.ols(formula=model5_linear_form, data=pokeaman_train)
model5_fit = model5_spec.fit()
model5_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.392
Model:                            OLS   Adj. R-squared:                  0.313
Method:                 Least Squares   F-statistic:                     4.948
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           9.48e-19
Time:                        03:28:45   Log-Likelihood:                -1765.0
No. Observations:                 400   AIC:                             3624.
Df Residuals:                     353   BIC:                             3812.
Df Model:                          46                                         
Covariance Type:            nonrobust                                         
==============================================================================================
                                 coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------
Intercept                     10.1046     14.957      0.676      0.500     -19.312      39.521
Legendary[T.True]             -3.2717      4.943     -0.662      0.508     -12.992       6.449
C(Generation)[T.2]             9.2938      4.015      2.315      0.021       1.398      17.189
C(Generation)[T.3]             2.3150      3.915      0.591      0.555      -5.385      10.015
C(Generation)[T.4]             4.8353      4.149      1.165      0.245      -3.325      12.995
C(Generation)[T.5]            11.4838      3.960      2.900      0.004       3.696      19.272
C(Generation)[T.6]             4.9206      4.746      1.037      0.300      -4.413      14.254
C(Q("Type 1"))[T.Dark]        -1.4155      6.936     -0.204      0.838     -15.057      12.226
C(Q("Type 1"))[T.Dragon]       0.8509      6.900      0.123      0.902     -12.720      14.422
C(Q("Type 1"))[T.Electric]    -6.3641      6.537     -0.974      0.331     -19.220       6.491
C(Q("Type 1"))[T.Fairy]       -1.9486     10.124     -0.192      0.847     -21.859      17.962
C(Q("Type 1"))[T.Fighting]     7.0308      7.432      0.946      0.345      -7.586      21.648
C(Q("Type 1"))[T.Fire]         3.0779      6.677      0.461      0.645     -10.055      16.210
C(Q("Type 1"))[T.Flying]      -2.1231     22.322     -0.095      0.924     -46.025      41.779
C(Q("Type 1"))[T.Ghost]        5.7343      8.488      0.676      0.500     -10.960      22.429
C(Q("Type 1"))[T.Grass]        3.3275      5.496      0.605      0.545      -7.481      14.136
C(Q("Type 1"))[T.Ground]       9.5118      7.076      1.344      0.180      -4.404      23.428
C(Q("Type 1"))[T.Ice]         -0.9313      7.717     -0.121      0.904     -16.108      14.246
C(Q("Type 1"))[T.Normal]      18.4816      5.312      3.479      0.001       8.034      28.929
C(Q("Type 1"))[T.Poison]       8.3411      7.735      1.078      0.282      -6.871      23.554
C(Q("Type 1"))[T.Psychic]      1.8061      6.164      0.293      0.770     -10.317      13.930
C(Q("Type 1"))[T.Rock]        -3.8558      6.503     -0.593      0.554     -16.645       8.933
C(Q("Type 1"))[T.Steel]       -4.0053      8.044     -0.498      0.619     -19.826      11.816
C(Q("Type 1"))[T.Water]        9.7988      5.166      1.897      0.059      -0.361      19.959
C(Q("Type 2"))[T.Dark]         5.8719     15.185      0.387      0.699     -23.993      35.737
C(Q("Type 2"))[T.Dragon]      13.2777     14.895      0.891      0.373     -16.016      42.571
C(Q("Type 2"))[T.Electric]    14.3228     17.314      0.827      0.409     -19.728      48.374
C(Q("Type 2"))[T.Fairy]        2.8426     14.268      0.199      0.842     -25.218      30.903
C(Q("Type 2"))[T.Fighting]     1.9741     14.089      0.140      0.889     -25.735      29.683
C(Q("Type 2"))[T.Fire]         0.2001     1

In [45]:
yhat_model5 = model5_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model5_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model5)[0,1]**2)

'In sample' R-squared:     0.3920134083531893
'Out of sample' R-squared: 0.30015614488652215


In [46]:
# Here's something a little more reasonable...
model6_linear_form = 'HP ~ Attack + Speed + Q("Sp. Def") + Q("Sp. Atk")'
# And here we'll add the significant indicators from the previous model
# https://chatgpt.com/share/81ab88df-4f07-49f9-a44a-de0cfd89c67c
model6_linear_form += ' + I(Q("Type 1")=="Normal")'
model6_linear_form += ' + I(Q("Type 1")=="Water")'
model6_linear_form += ' + I(Generation==2)'
model6_linear_form += ' + I(Generation==5)'

model6_spec = smf.ols(formula=model6_linear_form, data=pokeaman_train)
model6_fit = model6_spec.fit()
model6_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.333
Model:                            OLS   Adj. R-squared:                  0.319
Method:                 Least Squares   F-statistic:                     24.36
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           2.25e-30
Time:                        03:28:45   Log-Likelihood:                -1783.6
No. Observations:                 400   AIC:                             3585.
Df Residuals:                     391   BIC:                             3621.
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
======================================================================================================
                                         coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------
Intercept                             22.8587      3.876      5.897      0.000      15.238      30.479
I(Q("Type 1") == "Normal")[T.True]    17.5594      3.339      5.258      0.000      10.994      24.125
I(Q("Type 1") == "Water")[T.True]      9.0301      3.172      2.847      0.005       2.794      15.266
I(Generation == 2)[T.True]             6.5293      2.949      2.214      0.027       0.732      12.327
I(Generation == 5)[T.True]             8.4406      2.711      3.114      0.002       3.112      13.770
Attack                                 0.2454      0.037      6.639      0.000       0.173       0.318
Speed                                 -0.1370      0.045     -3.028      0.003      -0.226      -0.048
Q("Sp. Def")                           0.3002      0.045      6.662      0.000       0.212       0.389
Q("Sp. Atk")                           0.1192      0.042      2.828      0.005       0.036       0.202
==============================================================================
Omnibus:                      271.290   Durbin-Watson:                   1.999
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4238.692
Skew:                           2.651   Prob(JB):                         0.00
Kurtosis:                      18.040   Cond. No.                         618.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [47]:
yhat_model6 = model6_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model6_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2)

'In sample' R-squared:     0.3326310334310908
'Out of sample' R-squared: 0.29572460427079933


In [48]:
# And here's a slight change that seems to perhaps improve prediction...
model7_linear_form = 'HP ~ Attack * Speed * Q("Sp. Def") * Q("Sp. Atk")'
model7_linear_form += ' + I(Q("Type 1")=="Normal")'
model7_linear_form += ' + I(Q("Type 1")=="Water")'
model7_linear_form += ' + I(Generation==2)'
model7_linear_form += ' + I(Generation==5)'

model7_spec = smf.ols(formula=model7_linear_form, data=pokeaman_train)
model7_fit = model7_spec.fit()
model7_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.378
Model:                            OLS   Adj. R-squared:                  0.347
Method:                 Least Squares   F-statistic:                     12.16
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           4.20e-29
Time:                        03:28:46   Log-Likelihood:                -1769.5
No. Observations:                 400   AIC:                             3579.
Df Residuals:                     380   BIC:                             3659.
Df Model:                          19                                         
Covariance Type:            nonrobust                                         
==========================================================================================================
                                             coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------------
Intercept                                 95.1698     34.781      2.736      0.007      26.783     163.556
I(Q("Type 1") == "Normal")[T.True]        18.3653      3.373      5.445      0.000      11.733      24.997
I(Q("Type 1") == "Water")[T.True]          9.2913      3.140      2.959      0.003       3.117      15.466
I(Generation == 2)[T.True]                 7.0711      2.950      2.397      0.017       1.271      12.871
I(Generation == 5)[T.True]                 7.8557      2.687      2.923      0.004       2.572      13.140
Attack                                    -0.6975      0.458     -1.523      0.129      -1.598       0.203
Speed                                     -1.8147      0.554     -3.274      0.001      -2.905      -0.725
Attack:Speed                               0.0189      0.007      2.882      0.004       0.006       0.032
Q("Sp. Def")                              -0.5532      0.546     -1.013      0.312      -1.627       0.521
Attack:Q("Sp. Def")                        0.0090      0.007      1.311      0.191      -0.004       0.023
Speed:Q("Sp. Def")                         0.0208      0.008      2.571      0.011       0.005       0.037
Attack:Speed:Q("Sp. Def")                 -0.0002   9.06e-05     -2.277      0.023      -0.000   -2.82e-05
Q("Sp. Atk")                              -0.7277      0.506     -1.439      0.151      -1.722       0.267
Attack:Q("Sp. Atk")                        0.0136      0.005      2.682      0.008       0.004       0.024
Speed:Q("Sp. Atk")                         0.0146      0.007      2.139      0.033       0.001       0.028
Attack:Speed:Q("Sp. Atk")                 -0.0002    5.4e-05     -3.383      0.001      -0.000   -7.65e-05
Q("Sp. Def"):Q("Sp. Atk")                  0.0103      0.007      1.516      0.130      -0.003       0.024
Attack:Q("Sp. Def"):Q("Sp. Atk")          -0.0001   6.71e-05     -2.119      0.035      -0.000   -1.03e-05
Speed:Q("Sp. Def"):Q("Sp. Atk")           -0.0002   8.82e-05     -2.075      0.039      -0.000   -9.62e-06
Attack:Speed:Q("Sp. Def"):Q("Sp. Atk")   2.03e-06   7.42e-07      2.734      0.007     5.7e-07    3.49e-06
==============================================================================
Omnibus:                      252.300   Durbin-Watson:                   1.953
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3474.611
Skew:                           2.438   Prob(JB):                         0.00
Kurtosis:                      16.590   Cond. No.                     2.34e+09
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.34e+09. Thi

In [49]:
yhat_model7 = model7_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model7_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2)

'In sample' R-squared:     0.37818209127432456
'Out of sample' R-squared: 0.35055389205977444


In [50]:
# And here's a slight change that seems to perhas improve prediction...
model7_linear_form_CS = 'HP ~ scale(center(Attack)) * scale(center(Speed))'
model7_linear_form_CS += ' * scale(center(Q("Sp. Def"))) * scale(center(Q("Sp. Atk")))'
# We DO NOT center and scale indicator variables
model7_linear_form_CS += ' + I(Q("Type 1")=="Normal")'
model7_linear_form_CS += ' + I(Q("Type 1")=="Water")'
model7_linear_form_CS += ' + I(Generation==2)'
model7_linear_form_CS += ' + I(Generation==5)'

model7_CS_spec = smf.ols(formula=model7_linear_form_CS, data=pokeaman_train)
model7_CS_fit = model7_CS_spec.fit()
model7_CS_fit.summary().tables[-1] 
# "Cond. No." is NOW 15.4 due to centering and scaling

<class 'statsmodels.iolib.table.SimpleTable'>

In [51]:
# "Cond. No." WAS 2,340,000,000 WITHOUT to centering and scaling
model7_fit.summary().tables[-1]

<class 'statsmodels.iolib.table.SimpleTable'>

7) Model 5 expands from model 3 and 4 by the following:
In model 3, you initially just take into consideration attack and defense, then in model 4 you also take into account the speed, the legendary status and the special defense stats and their interactions with the other variables. However, in model 5, you remove the unnecessary interactions and only pay attention to the variables with likely more implication. In this case, you are taking into account attack, defense, speed, legendary status, special defense, special attack, and lastly the generation and type of the pokemon. In the fourth model, you took into consideration interactions between attack and everything else, which led into a lot of overlap. However, swapping the * signs out for + signs resulted in it to stop considering the implication and simply to account for the new data. This improves upon model 3 by adding more information, improves upon model 4 by reducing the number of unnecessary and overlapping interactions, and finally, improves upon both by taking into consideration the pokemon's generation.

7.2) A major downside of Model 5 was the generality in which it handled the data. Model 6 replaces the generation, legendary status, and types of the pokemon with simply focusing on whether or not the pokemon is Normal, Water, from Generation 2 or from Generation 5. It does this by setting them as indicator variables, where if the pokemon's main type is normal, it would satisfy the indicator variable for the normal type. This will result in a lot more nuance and specific data regarding the normal and water type pokemons, allowing you to tell the difference between normal, water, and every other pokemon type. It further lets you take into account the differences between generation 2, generation 5, and those belonging to a generation that isnt generation 2 or 5.

7.3) in the case of model 7, model 7 reimplements a lot of the interaction terms which were removed previously going from Model 4 to Model 5. It adds interactions between each of the variables yet again in a way that makes a bit more sense than in Model 4. This enables the model to determine the relationships between each of the predictor variables and likewise, the HP variable (dependent variable). The conditional number in model 7 after scaling and centering ends up becoming as low as 15.4, down from the ridiculously high e+16 numbers that Model 4 had. This means that the model can accurately determine the effect of each of the variables with a much higher certainty and less uncertainty between what each predictor variable's impact might be.

The rationale between each of these changes was to focus on specific categories, narrowing each of the scopes of what you were looking at. In the first few models, you were just looking at the general interactions between each of the pokeamans. However, as you reach the later points, you see narrowing of the scope down to comparing the difference between pokemon types, pokemon generations, and the like. 

In [52]:
import plotly.express as px  # etc.

songs_training_data,songs_testing_data = train_test_split(songs, train_size=31)
linear_form = 'danceability ~ energy * loudness + energy * mode'
   
reps = 100
in_sample_Rsquared = np.array([0.0]*reps)
out_of_sample_Rsquared = np.array([0.0]*reps)
for i in range(reps):
    songs_training_data,songs_testing_data = \
      train_test_split(songs, train_size=31)
    final_model_fit = smf.ols(formula=linear_form, 
                              data=songs_training_data).fit()
    in_sample_Rsquared[i] = final_model_fit.rsquared
    out_of_sample_Rsquared[i] = \
      np.corrcoef(songs_testing_data.danceability, 
                  final_model_fit.predict(songs_testing_data))[0,1]**2
    
df = pd.DataFrame({"In Sample Performance (Rsquared)": in_sample_Rsquared,
                   "Out of Sample Performance (Rsquared)": out_of_sample_Rsquared})     
fig = px.scatter(df, x="In Sample Performance (Rsquared)", 
                     y="Out of Sample Performance (Rsquared)")
fig.add_trace(go.Scatter(x=[0,1], y=[0,1], name="y=x", line_shape='linear'))

8) The way this works is by splitting the data provided into two sets, the training and testing sets. They will each account for 50% of the data provided. A linear model is fit to the training data each time using the code, indicated by the red line. Anything that sits near the line indicates that the model has similar performance between the training dataset and the testing dataset. If it's above the line, it could mean that the line is overfitted to the training data. Otherwise, it could mean that it is underfitting and not capturing necessary patterns. In the example shown here, it appears that the model is underfitting relatively consistently, as after running this about 30 times, the data is repeatedly "under" more often. However, in some variations, it is a relatively equal split and as such, it could be safe to assume that the model is relatively accurate. 

In [53]:
model7_gen1_predict_future = smf.ols(formula=model7_linear_form,
                                   data=pokeaman[pokeaman.Generation==1])
model7_gen1_predict_future_fit = model7_gen1_predict_future.fit()
print("'In sample' R-squared:    ", model7_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model7_gen1_predict_future_fit.rsquared, "(gen1_predict_future)")
y = pokeaman[pokeaman.Generation!=1].HP
yhat = model7_gen1_predict_future_fit.predict(pokeaman[pokeaman.Generation!=1])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1_predict_future)")

'In sample' R-squared:     0.37818209127432456 (original)
'Out of sample' R-squared: 0.35055389205977444 (original)
'In sample' R-squared:     0.5726118179916575 (gen1_predict_future)
'Out of sample' R-squared: 0.11151363354803218 (gen1_predict_future)


In [54]:
model7_gen1to5_predict_future = smf.ols(formula=model7_linear_form,
                                   data=pokeaman[pokeaman.Generation!=6])
model7_gen1to5_predict_future_fit = model7_gen1to5_predict_future.fit()
print("'In sample' R-squared:    ", model7_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model7_gen1to5_predict_future_fit.rsquared, "(gen1to5_predict_future)")
y = pokeaman[pokeaman.Generation==6].HP
yhat = model7_gen1to5_predict_future_fit.predict(pokeaman[pokeaman.Generation==6])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1to5_predict_future)")

'In sample' R-squared:     0.37818209127432456 (original)
'Out of sample' R-squared: 0.35055389205977444 (original)
'In sample' R-squared:     0.3904756578094535 (gen1to5_predict_future)
'Out of sample' R-squared: 0.23394915464343125 (gen1to5_predict_future)


In [55]:
model6_gen1_predict_future = smf.ols(formula=model6_linear_form,
                                   data=pokeaman[pokeaman.Generation==1])
model6_gen1_predict_future_fit = model6_gen1_predict_future.fit()
print("'In sample' R-squared:    ", model6_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model6_gen1_predict_future_fit.rsquared, "(gen1_predict_future)")
y = pokeaman[pokeaman.Generation!=1].HP
yhat = model6_gen1_predict_future_fit.predict(pokeaman[pokeaman.Generation!=1])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1_predict_future)")

'In sample' R-squared:     0.3326310334310908 (original)
'Out of sample' R-squared: 0.29572460427079933 (original)
'In sample' R-squared:     0.4433880517727282 (gen1_predict_future)
'Out of sample' R-squared: 0.1932858534276128 (gen1_predict_future)


In [56]:
model6_gen1to5_predict_future = smf.ols(formula=model6_linear_form,
                                   data=pokeaman[pokeaman.Generation!=6])
model6_gen1to5_predict_future_fit = model6_gen1to5_predict_future.fit()
print("'In sample' R-squared:    ", model6_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model6_gen1to5_predict_future_fit.rsquared, "(gen1to5_predict_future)")
y = pokeaman[pokeaman.Generation==6].HP
yhat = model6_gen1to5_predict_future_fit.predict(pokeaman[pokeaman.Generation==6])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1to5_predict_future)")

'In sample' R-squared:     0.3326310334310908 (original)
'Out of sample' R-squared: 0.29572460427079933 (original)
'In sample' R-squared:     0.33517279824114776 (gen1to5_predict_future)
'Out of sample' R-squared: 0.26262690178799936 (gen1to5_predict_future)


9) The model here illustrates that when using Model 7, in this case specifically talking about the "original" dataset provided to it, the in sample R squared is relatively good at 0.378, meaning that it can account for 37.8% of the variation in the model. However, the out of sample R squared is also quite high in this, at 0.350. This means that the model is relatively good at generalizing the data in relation to the model of linear regression made. If this number were higher, it would indicate that it is better at generalizing and would imply that the level of multicollinearity is relatively low, as the model is good at determining what impact each variable has on the overall result. When using generation 1 data, the in sample R squared is very high, this is to be expected as Model 7 was trained on generation 1 data, with indicator variables focused specifically on that. As a result, the in sample R squared is super high at 0.573. However, when acounting for the out of sample R squared, we find that it sits closer to 0.111, which means that it accounts for 11% of the variation. This suggests that in relation to the training set, which is what In Sample is accustomed to, the model performs very well. However, it is very bad when it comes to taking things it hasn't seen before and applying it like so. This could suggest that the multicollinearity for this model is relatively high and as such, the model is worse at generalization. A similar result is found when it is fed data from generation 1 to 5, in order to predict generation 6. The in sample R squared is relatively good at just under 35% of the variation being explained, but when given the task of generalizing the data, it struggles, resulting an out of sample R squared of 0.234, meaning it accounts for 23% of the variation.

Model 6 experiences a similar result, however, its out of sample R squared is consistently higher than the out of sample R squared faced in Model 7. However, alongside that, the in sample R squared for predicting the future is generally lower, never really surpassing 0.45, which means that Model 7, being more complex, is better at accounting for the specifics of the data set it is trained on. However, when given the task of generalizing new data, Model 7 has been overtrained on the data that it has been provided, and is too accustomed to the nuances such that when given a new task with new data, it is slightly worse at generalizing. Model 6 has less nuance and is simpler overall, which means that when given new data, it is relatively good at adapting. ChatGPT worded it in a good way in which model 7 is likely to overfit due to learning "idiosyncratic" (patterns that really have no correlation and are likely coincidence) patterns in the data which should not be applied to the larger dataset.

## ChatGPT Summary

In this conversation, we discussed the evaluation and interpretation of models in the context of predicting Pokémon stats across different generations. We focused on comparing the performance of two models (Model 6 and Model 7) using in-sample and out-of-sample R-squared values. Model 7, which is more complex and involves several interaction terms, initially showed higher in-sample R-squared values, indicating a better fit to the training data. However, when we tested its ability to predict future data, particularly for Generation 1, the model's out-of-sample R-squared was significantly lower, suggesting poor generalizability. This drop in performance is attributed to overfitting, where the complexity of Model 7 led it to capture noise and idiosyncratic patterns specific to the training data, which did not hold when applied to new, unseen data. In contrast, simpler models like Model 6, with fewer interaction terms, displayed more robust out-of-sample performance and generally showed stronger evidence for their coefficients, making them potentially more reliable and interpretable. The conversation highlighted the trade-off between model complexity and interpretability, emphasizing that simpler models often offer better generalizability and are easier to interpret, particularly when performance differences between models are not substantial.

https://chatgpt.com/share/6736bfbb-fa88-8002-a19b-1aa49629d23e